src/libYARP_dev/include/yarp/dev/IFrameSource.h

```diff
@@ -0,0 +1,200 @@
+    yarp::math::FrameTransform nullframe;
+public:
+    const auto begin() const { return storage.begin(); }
+    const auto end() const { return storage.end(); }
+    bool insertUpdate(const yarp::math::FrameTransform& frame)
+    {
+        storage[frame.frameId] = frame;
+        return true;
+    }
+
+    robotology::Result<yarp::math::FrameTransform> get(const std::string& frameid)
+    {
+        if (storage.find(frameid) == storage.end())
+        {
+            return { false, 0, "", nullframe };
+        }
+        return { true, 0, "", storage[frameid] };
+    }
+};
+
+class YARP_dev_API yarp::dev::IFrameSource
+{
+    enum ConnectionType
+    {
+        DIRECT,
+        INVERSE,
+        UNDIRECT,
+        DISCONNECTED
+    };
+    ConnectionType getConnectionType(const std::string& target_frame, const std::string& source_frame, std::string* commonAncestor = nullptr);
+
+    robotology::Result<yarp::math::FrameTransform> getChainedTransform(const std::string& target_frame_id, const std::string& source_frame_id);
+
+    virtual bool frameExistsRaw(const std::string &frame_id);
+
+    FrameEditor frameContainer;
+
+    virtual robotology::Result<std::string> getParentRaw(const std::string& frame_id);
+
+protected:
+    std::function<bool(IFrameSource*)> callback;
+    FrameEditor* getFrameContainer();
+    virtual void updateFrameContainer(FrameEditor& frameContainer) = 0;
+    virtual bool callbackPrepare() = 0;
+    virtual bool callbackDismiss() = 0;
+
+    
+public:
+    enum errors
+    {
+        OK = 0,
+        FRAME_NOT_FOUND,
+        TARGET_NOT_FOUND,
+        SOURCE_NOT_FOUND,
+        FRAMES_NOT_CONNECTED,
+        WRONG_INPUT_FORMAT,
+        TIMEOUT,
+        CALLBACK_ACTIVE,
+        UNDEFINED
+    };
+    virtual ~IFrameSource();
+
+    bool unsetCallback();
+
+    bool setCallback(std::function<bool(IFrameSource*)> cb);
+    /**
+    Creates a debug string containing the list of all registered frames.
+    * @param all_frames the returned string containing the frames
+    * @return std::string
+    */
+    virtual std::string allFramesAsString();
+
+    /**
+    Test if a transform exists.
+    * @param target_frame_id the name of target reference frame
+    * @param source_frame_id the name of source reference frame
+    * @return true/false
+    */
+    virtual robotology::Result<bool> canTransform(const std::string &target_frame, const std::string &source_frame);
+
+    /**
+    Check if a frame exists.
+    * @param frame_id the frame to be searched
+    * @param target_frame_id the name of target reference frame
+    * @param source_frame_id the name of source reference frame
+    * @return true/false
+    */
+    virtual bool frameExists(const std::string &frame_id);
+
+    /**
+    Gets a vector containing all the registered frames.
+    * @param ids the returned vector containing all frame ids
+    * @return true/false
+    */
+    virtual std::unordered_set<std::string> getAllFrameIds();
```

Comment:
For this and other methods, the old interface was: 
~~~
bool getAllFrameIds (std::vector< std::string > &ids) override
~~~
the old method permitted to get the the `frameIds` and the similar quantities without any dynamic memory allocation, while the new signature does not permit this. I suggest to also reintroduce the old signature to not constrain any user of the interface to have a dynamic memory allocation when using the interface.

Reply:
> the old method had also a dynamic memory allocation as the frame count where subject to change

As long as the capacity of the passed container is always greater than the maximum size, there is no dynamic memory allocation.  

>  we can simply make these container class member and return const references to them

How do you enforce proper mutual exclusion in this case?

Reply:
there's no way for the user to know the size currently and for the mutual exclusion we can making it thread-local.

Reply:
> there's no way for the user to know the size currently

In practice on a given closed system you can measure the maximum number of frames, and call `.reserve` with a value greater than that. More practically, what typically happens is that the vector is resized a few time, but as the capacity is not change when the vector if shrinked, so there actually are dynamic memory allocation only when the number of frames increased (not always to be honest, as typically the strategy used in `std::vector` on `resize` to increase its buffer is always to double its capacity if a reallocation is needed, regardless of the actual size requested. 

> perhaps we can making it thread-local

Personal opinion: I do not think that forcing any implementation of the `getAllFrameIds` method to have a thread local `std::vector<std::string>` buffer (or even just the additional complexity of doing it once) just to avoid the signature with the out variable passed as a reference is worth it. 

>  after a quick search i found this article https://web.archive.org/web/20130930101140/http://cpp-next.com/archive/2009/08/want-speed-pass-by-value
didn't read it completely but seems to definitely promote the pass by value practice for best performances and for some caveats it propose different solution from the classical output function argument way. i'll read it with attention asap.

That article  assumes that the in any case the `std::vector<std::string>` needs to be created ex-novo every time by the caller, and it is just arguing that for **non-virtual methods defined in the same compilation unit of caller** the compiler is able to ensure that the `std::vector<std::string>` is created only once. Our case is quite different: we are discussing **virtual  methods** (so the compiler at compile time has **no way** of  knowing which method will be actually called) and we are  discussing how to avoid any kind of memory allocation, i.e. that the `std::vector<std::string>` is  created even only once. 

> A good idea for get rid of this question once for all would be to do some test. i think i will do it.

Are you discussing **speed tests** or **dynamic memory allocation** tests? I  am actually arguing to have an interface that does not force dynamic memory allocation, the increased speed would be just a byproduct from my point of view.

> A thing that we can try is also to move the implementation of these method in the header in order to move the compilation of these function on the user side, allowing for further optimization in case of inner-loops calls.

As this is a virtual call that the users will call by just having a pointer to `IFrameSource`, no matter what, the method cannot be inlined by the compiler, because the compiler has no idea of which method is actually called.

Reply:
To be honest I just realized that on an increase in the number of the frames there always be a dynamic memory allocation if the additional frame id does not fit in the sso buffer. However not having dynamic memory allocation if the frame numbers do not change seems already quite convenient to me.

Reply:
> Personal opinion: I do not think that forcing any implementation of the getAllFrameIds method to have a thread local std::vector<std::string> buffer (or even just the additional complexity of doing it once) just to avoid the signature with the out variable passed as a reference is worth it.

i would personally get rid of this kind of signature cause it bothers me a lot. none of the framework i use has this kind of signature. this should mean something. anyway **if we find a way to have the same behaviour** with a signature semantically appropriate why don't use every c++11/14 construct to do it?

> Our case is quite different: we are discussing virtual methods (so the compiler at compile time has no way of knowing which method will be actually called)

this is true.. as it is true that is not mandatory at all for those method to be virtual. i had this doubt also for other reasons. removing the "virtualness" of the method would make the IFrameSource not an interface anymore, but a simple partially virtual public superclass. i don't know, i have to think about it

> Are you discussing speed tests or dynamic memory allocation tests? I am actually arguing to have an interface that does not force dynamic memory allocation, the increased speed would be just a byproduct from my point of view.

ok this was not clear for me, thank you for pointing it out. if we add the external reference to be valorized, the interface should do a check on the capacity of the vector and return fail if inappropriate right? in this case i should also add a method to know the frame count. but what to do with device such as the vicon, where new marker can pop in or out in any moment?

Reply:
>  anyway **if we find a way to have the same behaviour** with a signature semantically appropriate why don't use every c++11/14 construct to do it?

Not sure if I got what you mean here. 

> ok this was not clear for me, thank you for pointing it out. if we add the external reference to be valorized, the interface should do a check on the capacity of the vector and return fail if inappropriate right? in this case i should also add a method to know the frame count. but what to do with device such as the vicon, where new marker can pop in or out in any moment?

Consistently with the rest of YARP, I do not think we need to ensure there are no dynamic memory allocations at all, but I just think we should avoid to have an interface that forces the users to have a quite huge number of dynamic memory allocations.